### DISCUSSION TITLE
v2 Meta API

### DISCUSSION THREAD

**Discussion by 58bits** [meza/trance-stack#77] [Ideas]
If you're interested in implementing V2 meta - the original discussion is here...

https://github.com/remix-run/remix/discussions/4462

I've also implemented this here on our own 'remix workbench' site - with an example here...

https://github.com/infonomic/remix.infonomic.io/blob/develop/app/routes/_app%2B/notes.%24noteId_.edit.tsx

We also created helper utils for `mergeMeta` and `appendSiteTitle` which I _think_ are okay - here...

https://github.com/infonomic/remix.infonomic.io/blob/develop/app/utils/utils.ts 

Hope this helps...

**Comment by meza** [comment]
Very good shout, thanks! I didn't know about this. I think it's worth the upgrade as well as doing the route v2 restructuring as 1.14.1 already warns about the folder based solution's deprecation.

**Comment by 58bits** [comment]
Flat routes are really nice. They're also implemented in the repo above - although I've used https://github.com/kiliman/remix-flat-routes - which Michael kindly helped me with. The 'core' of Michael's repo is now in Remix, but I believe there are still a few features that are unique to Michael's repo - which he is going to continue to maintain.

**Reply by meza** [reply]
Feel free to send some PRs our way. I'm pretty tied down with the cookie policy vs. XSS protection topic rn, wouldn't want to dive into anything until I get to the end of that

**Reply by 58bits** [reply]
Would be glad too! Will be away this weekend but will send something your way next week ;-)

**Comment by 58bits** [comment]
Hey there - quick note to say I've not forgotten about this. Things went a bit bananas at this end. I think the V2 meta would be nice to start with. I'll try to get a PR over soon.

**Reply by meza** [reply]
No worries, I've been distracted big time too. Hopefully back on track from next week also

**Comment by 58bits** [comment]
Okay so I finally found a few minutes to take a stab at the V2 meta changes, as well as wrote the 'bare minimum' unit test required for our metaMerge util. I've cloned the main project, and while this isn't 1:1 based on your readme, we do have a developer Auth0 account, and I've updated the .env settings - at least the ones that I'd hoped would let me run this locally - but I'm seeing the following error after starting the app with `npm run dev` ... 

<img width="552" alt="image" src="https://user-images.githubusercontent.com/405612/230551029-9e943475-6d17-4bc4-b95c-fee08d9b10ae.png">

**Reply by meza** [reply]
If I were to guess it's probably missing the posthog integration for the feature flags?

**Comment by 58bits** [comment]
Tried a couple of things - sigh. Apologies, but struggling to get a development environment working. Also this just landed - https://github.com/remix-run/remix/pull/5785 - which means I'll have some changes to make when this is released. When they're in I'll post an update here with up-to-date links to my `metaMerge` util helper, and the updated meta functions ;-). More in a little while.

Oh and in the updates, there will be a link in the docs to a Gist that Ryan created for an alternative (and likely better?) mergeMeta helper method... https://gist.github.com/ryanflorence/ec1849c6d690cfbffcb408ecd633e069

**Comment by 58bits** [comment]
The docs are also much improved and summarize all changes well - including whether you need to merge meta or not...
https://remix.run/docs/en/main/route/meta-v2

**Comment by 58bits** [comment]
I've just updated our public project. Here's how it looks...

1. There's a new .env setting SITE_TITLE='Your Site' - which is appended to all title / og:title elements when present in mergeMeta
2. The `mergeMeta` and `appendSiteTitle`  utils are here...  https://github.com/infonomic/remix.infonomic.io/blob/develop/app/utils/utils.ts with very light test here  https://github.com/infonomic/remix.infonomic.io/blob/develop/app/utils/utils.test.ts
3. Two example routes here...  https://github.com/infonomic/remix.infonomic.io/blob/develop/app/routes/_app%2B/notes.%24noteId.tsx and https://github.com/infonomic/remix.infonomic.io/blob/develop/app/routes/_app%2B/notes.index.tsx

Hope some of this helps ;-)
